File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        '''
        # Solved during DSA Session 7:
        slow = head
        fast = head

        is_cycle_present = False

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

            if slow == fast:
                is_cycle_present = True
                break
            
        if is_cycle_present == False:
            return None

        slow = head
        cyclic_node = fast

        # Find the first node of the cycle;
        while slow != fast:
            slow = slow.next
            fast = fast.next
            cyclic_node = fast
        
        return cyclic_node
        '''

        # Solved during DSA Session 7:
        slow = head
        fast = head

        is_cycle_present = False

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

            if slow == fast:
                is_cycle_present = True
                break
            
        if is_cycle_present == False:
            return None

        slow = head

        # Find the first node of the cycle;
        while slow != fast:
            slow = slow.next
            fast = fast.next
        
        return fast

        

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        ListNode slowNode = head;
		ListNode fastNode = head;
		
		while(fastNode != null && fastNode.next != null){
			slowNode = slowNode.next;
			fastNode = fastNode.next.next;
			
			if(slowNode == fastNode){
				break;
			}
		}
		
		if(fastNode == null || fastNode.next == null){
			return null;
		}
		
		slowNode = head;
		
		while(slowNode != fastNode){
			slowNode = slowNode.next;
			fastNode = fastNode.next;
		}
		
		return slowNode;
    }
}